core/src/find_packet_sender_stake_stage.rs

```diff
@@ -0,0 +1,97 @@
+use {
+    crate::cluster_nodes::ClusterNodesCache,
+    crossbeam_channel::{Receiver, RecvTimeoutError, Sender},
+    rayon::{prelude::*, ThreadPool},
+    solana_gossip::cluster_info::ClusterInfo,
+    solana_perf::packet::PacketBatch,
+    solana_rayon_threadlimit::get_thread_count,
+    solana_runtime::bank_forks::BankForks,
+    solana_streamer::streamer::{self, StreamerError},
+    std::{
+        cell::RefCell,
+        collections::HashMap,
+        net::IpAddr,
+        sync::{Arc, RwLock},
+        thread::{self, Builder, JoinHandle},
+        time::{Duration, Instant},
+    },
+};
+
+const CLUSTER_NODES_CACHE_NUM_EPOCH_CAP: usize = 8;
+const CLUSTER_NODES_CACHE_TTL: Duration = Duration::from_secs(5);
+const STAKES_REFRESH_PERIOD_IN_MS: u128 = 1000;
+
+thread_local!(static PAR_THREAD_POOL: RefCell<ThreadPool> = RefCell::new(rayon::ThreadPoolBuilder::new()
+                    .num_threads(get_thread_count())
+                    .thread_name(|ix| format!("transaction_sender_stake_stage_{}", ix))
+                    .build()
+                    .unwrap()));
+
+pub struct FindPacketSenderStakeStage {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl FindPacketSenderStakeStage {
+    pub fn new(
+        packet_receiver: Receiver<PacketBatch>,
+        sender: Sender<Vec<PacketBatch>>,
+        bank_forks: Arc<RwLock<BankForks>>,
+        cluster_info: Arc<ClusterInfo>,
+    ) -> Self {
+        let cluster_nodes_cache = Arc::new(ClusterNodesCache::<FindPacketSenderStakeStage>::new(
+            CLUSTER_NODES_CACHE_NUM_EPOCH_CAP,
+            CLUSTER_NODES_CACHE_TTL,
+        ));
+        let thread_hdl = Builder::new()
+            .name("sol-tx-sender_stake".to_string())
+            .spawn(move || {
+                let mut last_stakes = Instant::now();
+                let mut ip_to_stake: HashMap<IpAddr, u64> = HashMap::new();
+                loop {
+                    if last_stakes.elapsed().as_millis() > STAKES_REFRESH_PERIOD_IN_MS {
+                        let (root_bank, working_bank) = {
+                            let bank_forks = bank_forks.read().unwrap();
+                            (bank_forks.root_bank(), bank_forks.working_bank())
+                        };
+                        ip_to_stake = cluster_nodes_cache
+                            .get(root_bank.slot(), &root_bank, &working_bank, &cluster_info)
+                            .get_ip_to_stakes();
+                        last_stakes = Instant::now();
+                    }
```

Comment:
I think this loop may be unnecessary because the call to `cluster_nodes_cache.get()` will know to refresh every `CLUSTER_NODES_CACHE_TTL` (which was passed on construction).

Thus I think you could have to call to `get()` return a signal on whether the cache was refreshed, i.e. if this path is skipped: https://github.com/solana-labs/solana/blob/734a684c47ca05fd976783c9f3b2e0ce4503cdea/core/src/cluster_nodes.rs#L464

cc @behzadnouri to see if he has any comments about best practices in this code path. For context, Tao wants to maintain a mapping from `ip` to `stakes` for the validators

Reply:
This is for a future note (think we should create an issue to remember to address this), that when the QUIC stuff lands, this ip checking will become unnecessary. 

We will know for each QUIC connection which validator id the packets are arriving from, and then we will just need to check `bank.epoch_staked_nodes.get(validator_id)` to get the stake.

This means we will be able to rip out all of this `ClusterNodes` stuff

Reply:
Not a strong opinion, but feel `fn get(...) -> (nodes, cache_has_refreshed_as_bool) {...}` pollutes api a bit. If we w ant to do it, may as well add a separate function `fn is_cache_due_for_refresh()->bool`. wdyt @behzadnouri @carllin  

BTW, I think you mean `if last_stakes.elapsed().as_millis() > STAKES_REFRESH_PERIOD_IN_MS {` is unnecessary. So if I replace `STAKES_REFRESH_PERIOD_IN_MS` with `CLUSTER_NODES_CACHE_TTL`, then it will not unnecessarily call `get_ip_to_stakes()` when cache hasn't refreshed.

Reply:
yeah, agree. my comments above.